src/plugins/ucc_discovery_plugin.hpp

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include "expression/abstract_expression.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "utils/abstract_plugin.hpp"
+#include "utils/pausable_loop_thread.hpp"
+
+namespace opossum {
+
+class UCCCandidate {
+ public:
+  UCCCandidate(std::string table_name, ColumnID column_id) : _table_name(table_name), _column_id(column_id) {}
+
+  const std::string table_name() const {
+    return this->_table_name;
+  }
+
+  const ColumnID column_id() const {
+    return this->_column_id;
+  }
+
+  friend bool operator==(const UCCCandidate& lhs, const UCCCandidate& rhs) {
+    return (lhs.column_id() == rhs.column_id()) && (lhs.table_name() == rhs.table_name());
+  }
+
+ protected:
+  std::string _table_name;
+  ColumnID _column_id;
+};
+
+using UCCCandidates = std::unordered_set<UCCCandidate>;
+
+class UccDiscoveryPlugin : public AbstractPlugin {
+ public:
+  std::string description() const final;
+
+  void start() final;
+
+  void stop() final;
+
+  std::vector<std::pair<PluginFunctionName, PluginFunctionPointer>> provided_user_executable_functions() const final;
+
+ protected:
+  friend class UccDiscoveryPluginTest;
+
+  UCCCandidates identify_ucc_candidates() const;
+  std::shared_ptr<std::vector<UCCCandidate>> generate_valid_candidates(
```

Comment:
Should these internal functions be prefixed with an underscore?
```suggestion
  UCCCandidates _identify_ucc_candidates() const;
  std::shared_ptr<std::vector<UCCCandidate>> _generate_valid_candidates(
```